z3/src/optimize.rs

```diff
@@ -34,6 +35,37 @@ impl<'ctx> Optimize<'ctx> {
         unsafe { Z3_optimize_assert(self.ctx.z3_ctx, self.z3_opt, ast.get_z3_ast()) };
     }
 
+    /// Assert soft constraint to the optimization context.
+    /// Weight is a rational number penalty for violating the constraint.
+    /// Group is an optional identifier to group soft constraints.
+    ///
+    /// # See also:
+    ///
+    /// - [`Optimize::maximize()`](#method.maximize)
+    /// - [`Optimize::minimize()`](#method.minimize)
+    pub fn assert_soft<W: ToString>(
+        &self,
+        ast: &impl Ast<'ctx>,
+        weight: W,
```

Comment:
The `W: ToString` bound seems overly permissive here. Am I understanding correctly that it should be allowed to be either an integer, a float, or some kind of a rational number string like `"1 / 2"`?

I think we can tighten this up so that users can't pass in things like `"blue hippopotamus"` -- resulting in who knows what behavior from Z3 -- by defining a type and a few conversion impls:

```rust
pub enum Weight {
    Int(i64),
    Float(f64),
    Rational(i64, i64),
}

// i64 -> Weight::Int
impl From<i64> for Weight { /* ... */ }

// f64 -> Weight::Float
impl From<f64> for Weight { /* ... */ }

// (i64, i64) -> Weight::Rational
impl From<(i64, i64)> for Weight { /* ... */ }
```

And then instead of accepting any `W: ToString`, this function would accept any `W: Into<Weight>`, do `w.into()` to convert it into a weight, and then match on which kind of weight it is and format it into a C string as expected by Z3.

Reply:
Good point, I think that's a good idea. Will revise

Reply:
I'm looking at the function and it's unclear even to me, so I'll dig a bit deeper and figure out what exact weights are valid

Reply:
I settled on using the sealed trait pattern to implement to_string on all integer types and integer pairs (for rationals)